kubeadm: add 1.32 documentation

neolit123 opened the pull request:
PLACEHOLDER

neolit123 commented:
/sig cluster-lifecycle
/kind feature cleanup

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/kubeadm-init.md:
```diff
@@ -155,9 +155,9 @@ List of feature gates:
 {{< table caption="kubeadm feature gates" >}}
 Feature | Default | Alpha | Beta | GA
 :-------|:--------|:------|:-----|:----
+`NodeLocalCRISocket` | `false` | 1.32 | - | -
```
Keep an alphabeta order of these gates?

neolit123 commented on content/en/docs/reference/setup-tools/kubeadm/kubeadm-upgrade-phase.md:
```diff
@@ -1,23 +1,37 @@
-title: kubeadm upgrade phase
-weight: 90
+title: kubeadm upgrade phases
+weight: 40
```
making this appear after the `kubeadm upgrade` page in the list

neolit123 commented:
/cc @SataQiu @pacoxu @carlory 

i added some docs we missed adding when 1.31 (v1beta4) was added.
are we missing anything else for 1.32?

pacoxu commented:
/cc @HirazawaUi
for NodeLocalCRISocket

pacoxu commented:
https://kubernetes.io/docs/reference/setup-tools/kubeadm/implementation-details/#preflight-checks
Is this generated? Need we do some update on this page?

Some may needs an update: for instance, the crictl.
```
Starting from 1.24, kubeadm uses crictl to communicate to all known CRI endpoints.
```
https://kubernetes.io/docs/reference/setup-tools/kubeadm/implementation-details/#preflight-checks-1

pacoxu commented:
Maybe this one: https://github.com/kubernetes/kubernetes/pull/126032 
- a quick search for kubeadm features in milestone v1.32 https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+milestone%3Av1.32+is%3Aclosed+label%3Akind%2Ffeature+label%3Aarea%2Fkubeadm

neolit123 commented:
> https://kubernetes.io/docs/reference/setup-tools/kubeadm/implementation-details/#preflight-checks Is this generated? Need we do some update on this page?
> 
> Some may needs an update: for instance, the crictl.
> 
> ```
> Starting from 1.24, kubeadm uses crictl to communicate to all known CRI endpoints.
> ```
> 
> https://kubernetes.io/docs/reference/setup-tools/kubeadm/implementation-details/#preflight-checks-1

the implementation details need update, yes.

neolit123 commented:
> Maybe this one: [kubernetes/kubernetes#126032](https://github.com/kubernetes/kubernetes/pull/126032)
> 

upgrade apply phases are covered in the generated ref docs in this pr. but will check if another place needs update too.

> * a quick search for kubeadm features in milestone v1.32 https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+milestone%3Av1.32+is%3Aclosed+label%3Akind%2Ffeature+label%3Aarea%2Fkubeadm

will check.

pacoxu commented on content/en/docs/reference/setup-tools/kubeadm/kubeadm-init.md:
```diff
@@ -199,11 +202,18 @@ List of deprecated feature gates:
+`PublicKeysECDSA`
+: Can be used to create a cluster that uses ECDSA certificates instead of the default RSA algorithm.
+Renewal of existing ECDSA certificates is also supported using `kubeadm certs renew`, but you cannot
+switch between the RSA and ECDSA algorithms on the fly or during upgrades.Kubernetes versions before v1.31 had a bug where keys in generated kubeconfig files were set use RSA, even when you had enabled the `PublicKeysECDSA` feature gate. This feature gate is deprecated in favor of the `encryptionAlgorithm`
```
```suggestion
switch between the RSA and ECDSA algorithms on the fly or during upgrades.Kubernetes versions before
v1.31 had a bug where keys in generated kubeconfig files were set use RSA, even when you had enabled
the `PublicKeysECDSA` feature gate. This feature gate is deprecated in favor of the `encryptionAlgorithm`
```

neolit123 commented:
> overall lgtm.
> 
> Should we review the generated docs?

you can review them if you'd like. but the fixes must be done in k/k.

neolit123 commented:
updated.
applied certain fixes and also added some missing info in implementation-details.

sftim commented on content/en/docs/reference/setup-tools/kubeadm/kubeadm-init.md:
```diff
@@ -176,20 +176,23 @@ policy during rolling upgrades.
-despite the feature gate being enabled. Kubernetes versions before v1.31 had a bug where keys in generated kubeconfig files
-were set use RSA, even when you had enabled the `PublicKeysECDSA` feature gate.
+`NodeLocalCRISocket`
+: With this feature gate enabled, kubeadm will consume the CRI socket for each node from the file
```
What does “consume“ mean?
I guess it means that kubeadm parses that file and then works out what the path to the socket should be; have I got that right?

Ideally we document the format of this configuration file and update https://kubernetes.io/docs/reference/node/kubelet-files/ 

- if the kubelet does use that file, we should definitely update the reference
- if the kubelet ignores that file, but someone might find a file `/var/lib/kubelet/instance-config.yaml` on their node, we should give it a special mention within https://kubernetes.io/docs/reference/node/kubelet-files/  to explain the situation.

neolit123 replied:
> What does “consume“ mean?

consume == parse
i can change the verb.

> I guess it means that kubeadm parses that file and then works out what the path to the socket should be; have I got that right?

yes, since the file contents are a single line `containerRuntimeEndpoint: VALUE` i will add it in the FG text as an example.

> https://kubernetes.io/docs/reference/node/kubelet-files/

it's not a kubelet file and the kubelet does not know about it, so it should not be covered in that page.

neolit123 replied:
updated for more clarity.

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -76,6 +76,20 @@ in a majority of cases, and the most intuitive location; other constant paths an
   - `front-proxy-ca.crt`, `front-proxy-ca.key` for the front proxy certificate authority
   - `front-proxy-client.crt`, `front-proxy-client.key` for the front proxy client
 
+## The kubeadm configuration API
```
```suggestion
## The kubeadm configuration file format
```
We stretch the term “application programming interface“ really far. These are really configuration files that happen to align with our API conventions.

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -76,6 +76,20 @@ in a majority of cases, and the most intuitive location; other constant paths an
+
+Most kubeadm commands support a `--config` flag which allows passing a configuration file from
+disk. The configuration file format follows the common Kubernetes API `apiVersion` / `kind` scheme,
+but is considered a component configuration API, similarly to that of the kubelet.
```
```suggestion
but is considered a component configuration format. Several Kubernetes components, such as the kubelet,
also support file-based configuration.
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -76,6 +76,20 @@ in a majority of cases, and the most intuitive location; other constant paths an
+disk. The configuration file format follows the common Kubernetes API `apiVersion` / `kind` scheme,
+but is considered a component configuration API, similarly to that of the kubelet.
+
+Different kubeadm commands require a different `kind`. For example, `InitConfiguration` for `kubeadm init`, `JoinConfiguration` for `kubeadm join`, `UpgradeConfiguration` for `kubeadm upgrade` and `ResetConfiguration`
```
(nit)
```suggestion
Different kubeadm subcommands require a different `kind` of configuration file.
For example, `InitConfiguration` for `kubeadm init`, `JoinConfiguration` for `kubeadm join`, `UpgradeConfiguration` for `kubeadm upgrade` and `ResetConfiguration`
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -76,6 +76,20 @@ in a majority of cases, and the most intuitive location; other constant paths an
+Different kubeadm commands require a different `kind`. For example, `InitConfiguration` for `kubeadm init`, `JoinConfiguration` for `kubeadm join`, `UpgradeConfiguration` for `kubeadm upgrade` and `ResetConfiguration`
+for `kubeadm reset`.
+
+The command `kubeadm config migrate` can be used to migrate an older, deprecated API file
+to a newer API file.
```
Is this right?
```suggestion
The command `kubeadm config migrate` can be used to migrate an older format configuration
file to a newer (current) configuration format. The kubeadm tool only supports migrating from
deprecated configuration formats to the current format.
```
I reworded into the active voice per the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/).

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -223,6 +237,8 @@ The `super-admin.conf` file must be stored in a safe location and should not be
 
+To generate kubeconfig files for additional users the command `kubeadm kubeconfig user` can be used.
+
 Please note that:
```
Optional extra tweak:
```suggestion
Also note that:
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -223,6 +237,8 @@ The `super-admin.conf` file must be stored in a safe location and should not be
 See [RBAC user facing role bindings](/docs/reference/access-authn-authz/rbac/#user-facing-roles)
 for additional information on RBAC and built-in ClusterRoles and groups.
 
+To generate kubeconfig files for additional users the command `kubeadm kubeconfig user` can be used.
```
Does this look right?
```suggestion
You can run [`kubeadm kubeconfig user`](/docs/reference/setup-tools/kubeadm/kubeadm-kubeconfig/#cmd-kubeconfig-user)
to generate kubeconfig files for additional users.

{{< caution >}}
The generated configuration files include an embedded authentication key, and you should treat
them as confidential.
{{< /caution >}}
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -76,6 +76,20 @@ in a majority of cases, and the most intuitive location; other constant paths an
+The command `kubeadm config migrate` can be used to migrate an older, deprecated API file
+to a newer API file.
+
+See the [API reference](/docs/reference/config-api/kubeadm-config.v1beta4/) page for more details.
```
```suggestion
See the [kubeadm configuration reference](/docs/reference/config-api/kubeadm-config.v1beta4/) page for more details.
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+
+## kubeadm reset workflow internal design
+
+The command `kubeadm reset` can be used to reset a node where kubeadm commands previously executed.
```
```suggestion
You can use the `kubeadm reset` subcommand a node where kubeadm commands previously executed.
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+## kubeadm reset workflow internal design
+
+The command `kubeadm reset` can be used to reset a node where kubeadm commands previously executed.
+Please note that `kubeadm reset` is considered a best-effort cleanup of the node.
```
```suggestion
This subcommand performs a **best-effort** cleanup of the node.
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -548,9 +557,7 @@ preconditions and avoid common cluster startup problems.
 Please note that:
```
Optional extra rewording:
```suggestion
Also note that:
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+## kubeadm upgrade workflow internal design
+
+`kubeadm upgrade` has sub-commands for handling the upgrade of the Kubernets cluster created by kubeadm.
+`kubeadm upgrade apply` must be called on a control plane node, chosen by the user, to start
+the upgrade process and `kubeadm upgrade node` must be called on any remaining control plane or worker nodes.
```
```suggestion
Tou must run `kubeadm upgrade apply` on a control plane node (you can choose which one);
this starts the upgrade process. You then run run `kubeadm upgrade node` on all remaining
nodes (both worker nodes and control plane nodes).
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+`kubeadm upgrade apply` must be called on a control plane node, chosen by the user, to start
+the upgrade process and `kubeadm upgrade node` must be called on any remaining control plane or worker nodes.
+
+Both `kubeadm upgrade apply` and `kubeadm upgrade node` have a `phase` subcommand which grants access
```
(nit)
```suggestion
Both `kubeadm upgrade apply` and `kubeadm upgrade node` have a `phase` subcommand which provides access
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+
+### kubeadm upgrade plan
+
+Can be optionally called before `kubeadm upgrade apply`. It checks which versions are available to upgrade
```
```suggestion
You can optionally run `kubeadm upgrade plan` before you run `kubeadm upgrade apply`.
The `plan` subcommand checks which versions are available to upgrade
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+### kubeadm upgrade node
+
+`kubeadm upgrade node` upgrades a single control plane or worker node after the cluster upgrade has
+starteed with `kubeadm upgrade apply`. The command detects if the node is a control plane node by checking
```
```suggestion
started (by running `kubeadm upgrade apply`). The command detects if the node is a control plane node by checking
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+
+`kubeadm upgrade node` upgrades a single control plane or worker node after the cluster upgrade has
+starteed with `kubeadm upgrade apply`. The command detects if the node is a control plane node by checking
+if the file `/etc/kubernetes/manifests/kube-apiserver.yaml` exists. That means there is a running kube-apiserver
+on the node.
```
```suggestion
if the file `/etc/kubernetes/manifests/kube-apiserver.yaml` exists. On finding that file, the kubeadm tool
infers that there is a running kube-apiserver Pod on this node.
```
?

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+- For control plane nodes, upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests`
+and waits for the kubelet to restart the components if the files have changed.
+- Writes updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.
+- For control plane nodes, Upgrades the kube-proxy and CoreDNS addons conditionally if all existing
```
```suggestion
- (For control plane nodes) upgrades the kube-proxy and CoreDNS
  {{< glossary_tooltip text="addons" term_id="addons" >}} conditionally, provided that all existing
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+The command supports a configuration file.
+
+Additionally:
+- IPVS and iptables rules are not cleaned.
+- CNI configuration is not cleaned.
+- `.kube/` in the home user directory is not cleaned.
```
```suggestion
- IPVS, iptables and nftables rules are **not** cleaned up.
- CNI (network plugin) configuration is **not** cleaned up.
- `.kube/` in the user's home directory is **not** cleaned up.
```

sftim commented on content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:
````diff
@@ -45,6 +47,37 @@ kubeadm does not overwrite them. This means you can, for example, copy an existi
+```yaml
+apiVersion: kubeadm.k8s.io/v1beta4
+kind: ClusterConfiguration
+certificateValidityPeriod: 8760h # Default: 365 days * 24 hours = 1 year
+caCertificateValidityPeriod: 87600h # Default: 365 days * 24 hours * 10 = 10 years
````
(nit)
```suggestion
certificateValidityPeriod: 8760h # Default: 365 days × 24 hours = 1 year
caCertificateValidityPeriod: 87600h # Default: 365 days × 24 hours * 10 = 10 years
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +626,85 @@ is deleted.
+### kubeadm upgrade diff
+
+Shows what differences would be applied to existing static pod manifests for control plane nodes.
+A more verbose option to do the same is running `kubeadm upgrade apply --dry-run` or
```
```suggestion
A more verbose way to do the same thing is running `kubeadm upgrade apply --dry-run` or
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -582,9 +589,8 @@ In order to prevent "man in the middle" attacks, several steps are taken:
-
-Pub key validation can be skipped passing `--discovery-token-unsafe-skip-ca-verification` flag;
-This weakens the kubeadm security model since others can potentially impersonate the Kubernetes Master.
+CA validation can be skipped by passing `--discovery-token-unsafe-skip-ca-verification` flag;
```
```suggestion
You can skip CA validation by passing the `--discovery-token-unsafe-skip-ca-verification` flag on the command line.
```

sftim commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -253,7 +269,7 @@ Static Pod manifests share a set of common properties:
 - Leader election is enabled for both the controller-manager and the scheduler
 - Controller-manager and the scheduler will reference kubeconfig files with their respective, unique identities
-- All static Pods get any extra flags specified by the user as described in
+- All static Pods get any extra flags or patches specified by the user as described in
```
(nit / optional style improvement)
```suggestion
- All static Pods get any extra flags or patches that you specify, as described in
```

sftim reviewed (commented):
Thanks. I've made a mix of suggestions; some are just style nits, some are places where I didn't understand the doc (so, I assume other people also might not understand), and some are around saying "configuration file" (or "configuration file format") rather than "configuration API".

"Configuration API" might have some people thinking you talk HTTP and then PUT your changes, which isn't what we're describing.

neolit123 commented:
/label tide/merge-method-squash

tengqm commented:
@neolit123 I noticed that this PR comprises both manually created contents and generated contents. Can/shall we separate them into two PRs?

neolit123 commented:
/remove-label tide/merge-method-squash

neolit123 commented:
> @neolit123 I noticed that this PR comprises both manually created contents and generated contents. Can/shall we separate them into two PRs?

i split this PR in two commits:
- generated ref docs
- authored content updates

and removed the squash label.

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+## kubeadm upgrade workflow internal design
+
+`kubeadm upgrade` has sub-commands for handling the upgrade of the Kubernets cluster created by kubeadm.
+Tou must run `kubeadm upgrade apply` on a control plane node (you can choose which one);
```
```suggestion
You must run `kubeadm upgrade apply` on a control plane node (you can choose which one);
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+
+### kubeadm upgrade diff
+
+Shows what differences would be applied to existing static pod manifests for control plane nodes.
```
```suggestion
This shows what differences would be applied to existing static pod manifests for control plane nodes.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+
+### kubeadm upgrade apply
+
+`kubeadm upgrade apply` takes cane of preparing the cluster for the upgrade of all nodes, and also
```
```suggestion
`kubeadm upgrade apply` prepares the cluster for the upgrade of all nodes, and also
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+### kubeadm upgrade apply
+
+`kubeadm upgrade apply` takes cane of preparing the cluster for the upgrade of all nodes, and also
+upgrades the control plane node where its called. The steps it performs are:
```
```suggestion
upgrades the control plane node where it's run. The steps it performs are:
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+`kubeadm upgrade apply` takes cane of preparing the cluster for the upgrade of all nodes, and also
+upgrades the control plane node where its called. The steps it performs are:
+
+- Runs preflight checks similarly to `kubeadm init` and `kubeadm join` ensuring container images are downloaded
+and the cluster is in a good state to be upgraded.
```
```suggestion
- Runs preflight checks similarly to `kubeadm init` and `kubeadm join`, ensuring container images are downloaded
  and the cluster is in a good state to be upgraded.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+- Runs preflight checks similarly to `kubeadm init` and `kubeadm join` ensuring container images are downloaded
+and the cluster is in a good state to be upgraded.
+- Upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests` and waits
+for the kubelet to restart the components if the files have changed.
```
```suggestion
  for the kubelet to restart the components if the files have changed.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+and the cluster is in a good state to be upgraded.
+- Upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests` and waits
+for the kubelet to restart the components if the files have changed.
+- Uploads updated kubeadm and kubelet configurations to the cluster in the respective `kube-system/kubeadm-config`
```
```suggestion
- Uploads the updated kubeadm and kubelet configurations to the cluster in the `kubeadm-config`
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+- Upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests` and waits
+for the kubelet to restart the components if the files have changed.
+- Uploads updated kubeadm and kubelet configurations to the cluster in the respective `kube-system/kubeadm-config`
+and `kube-system/kubelet-config` ConfigMaps.
```
```suggestion
  and the `kubelet-config` ConfigMaps (both in the `kube-system` namespace).
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+- Uploads updated kubeadm and kubelet configurations to the cluster in the respective `kube-system/kubeadm-config`
+and `kube-system/kubelet-config` ConfigMaps.
+- Writes updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.
+- Configures bootstrap token and `kube-public/cluster-info` ConfigMap RBAC rules. This is the same as
```
```suggestion
- Configures bootstrap token and the `cluster-info` ConfigMap for RBAC rules. This is the same as
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+and `kube-system/kubelet-config` ConfigMaps.
+- Writes updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.
+- Configures bootstrap token and `kube-public/cluster-info` ConfigMap RBAC rules. This is the same as
+in the `kubeadm init` stage and ensures that the cluster continues to support nodes joining with bootstrap tokens.
```
```suggestion
  in the `kubeadm init` stage and ensures that the cluster continues to support nodes joining with bootstrap tokens.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+- Configures bootstrap token and `kube-public/cluster-info` ConfigMap RBAC rules. This is the same as
+in the `kubeadm init` stage and ensures that the cluster continues to support nodes joining with bootstrap tokens.
+- Upgrades the kube-proxy and CoreDNS addons conditionally if all existing kube-apiservers in the cluster
+have already been upgraded to a the target version.
```
```suggestion
  have already been upgraded to the target version.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+in the `kubeadm init` stage and ensures that the cluster continues to support nodes joining with bootstrap tokens.
+- Upgrades the kube-proxy and CoreDNS addons conditionally if all existing kube-apiservers in the cluster
+have already been upgraded to a the target version.
+- Performs any post upgrade tasks, such as, release specific cleanups of deprecated features.
```
```suggestion
- Performs any post-upgrade tasks, such as, cleaning up deprecated features which are release specific.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+
+- Runs preflight checks similarly to `kubeadm upgrade apply`.
+- For control plane nodes, upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests`
+and waits for the kubelet to restart the components if the files have changed.
```
```suggestion
  and waits for the kubelet to restart the components if the files have changed.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+- Runs preflight checks similarly to `kubeadm upgrade apply`.
+- For control plane nodes, upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests`
+and waits for the kubelet to restart the components if the files have changed.
+- Writes updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.
```
```suggestion
- Writes the updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+- Writes updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.
+- (For control plane nodes) upgrades the kube-proxy and CoreDNS
+  {{< glossary_tooltip text="addons" term_id="addons" >}} conditionally, provided that all existing
+kube-apiservers in the cluster have already been upgraded to a the target version.
```
```suggestion
  API servers in the cluster have already been upgraded to the target version.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+- (For control plane nodes) upgrades the kube-proxy and CoreDNS
+  {{< glossary_tooltip text="addons" term_id="addons" >}} conditionally, provided that all existing
+kube-apiservers in the cluster have already been upgraded to a the target version.
+- Performs any post upgrade tasks, such as, release specific cleanups of deprecated features.
```
```suggestion
- Performs any post-upgrade tasks, such as cleaning up deprecated features which are release specific.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+
+## kubeadm reset workflow internal design
+
+You can use the `kubeadm reset` subcommand a node where kubeadm commands previously executed.
```
```suggestion
You can use the `kubeadm reset` subcommand on a node where kubeadm commands previously executed.
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/kubeadm-init.md:
```diff
@@ -176,20 +176,24 @@ policy during rolling upgrades.
+: With this feature gate enabled, kubeadm will read/write the CRI socket for each node from/to the file
+`/var/lib/kubelet/instance-config.yaml` instead of reading/writing it from/to the annotation
+`kubeadm.alpha.kubernetes.io/cri-socket` on the Node object. The new file is applied as an instance
+configuration patch, before any other user managed patches are applied when the `--patches` functionality
```
```suggestion
configuration patch, before any other user managed patches are applied when the `--patches` flag
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/kubeadm-upgrade-phase.md:
```diff
@@ -1,23 +1,37 @@
+
+## kubeadm upgrade apply phase {#cmd-apply-phase}
+
+Using the phases of `kubeadm upgrade apply` you can choose to execute the separate steps of the initial upgrade
```
```suggestion
Using the phases of `kubeadm upgrade apply`, you can choose to execute the separate steps of the initial upgrade
```

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/kubeadm-upgrade-phase.md:
```diff
@@ -1,23 +1,37 @@
+Using the phases of `kubeadm upgrade apply` you can choose to execute the separate steps of the initial upgrade
+of a control plane node.
+
+{{< tabs name="tab-phase" >}}
+{{< tab name="phase" include="generated/kubeadm_upgrade/kubeadm_upgrade_apply_phase.md" />}}
+{{< tab name="preflight" include="generated/kubeadm_upgrade/kubeadm_upgrade_apply_phase_preflight.md" />}}
+{{< tab name="control-plane" include="generated/kubeadm_upgrade/kubeadm_upgrade_apply_phase_control-plane.md" />}}
+{{< tab name="upload-config" include="generated/kubeadm_upgrade/kubeadm_upgrade_apply_phase_upload-config.md" />}}
+{{< tab name="kubelet-config" include="generated/kubeadm_upgrade/kubeadm_upgrade_apply_phase_kubelet-config.md" />}}
+{{< tab name="bootstrap-token" include="generated/kubeadm_upgrade/kubeadm_upgrade_apply_phase_bootstrap-token.md" />}}
+{{< tab name="addon" include="generated/kubeadm_upgrade/kubeadm_upgrade_apply_phase_addon.md" />}}
+{{< tab name="post-upgrade" include="generated/kubeadm_upgrade/kubeadm_upgrade_apply_phase_post-upgrade.md" />}}
+{{< /tabs >}}
```
I don't think tabs are the best option here.

tengqm commented on content/en/docs/reference/setup-tools/kubeadm/kubeadm-upgrade-phase.md:
```diff
@@ -1,23 +1,37 @@
 {{< tab name="kubelet-config" include="generated/kubeadm_upgrade/kubeadm_upgrade_node_phase_kubelet-config.md" />}}
+{{< tab name="addon" include="generated/kubeadm_upgrade/kubeadm_upgrade_node_phase_addon.md" />}}
+{{< tab name="post-upgrade" include="generated/kubeadm_upgrade/kubeadm_upgrade_node_phase_post-upgrade.md" />}}
 {{< /tabs >}}
```
Here too.
These phases are not mutual exclusive IIUC.

SataQiu commented on content/en/docs/reference/setup-tools/kubeadm/implementation-details.md:
```diff
@@ -620,3 +635,88 @@ is deleted.
+
+`kubeadm upgrade` has sub-commands for handling the upgrade of the Kubernets cluster created by kubeadm.
+You must run `kubeadm upgrade apply` on a control plane node (you can choose which one);
+this starts the upgrade process. You then run run `kubeadm upgrade node` on all remaining
```
```suggestion
this starts the upgrade process. You then run `kubeadm upgrade node` on all remaining
```

SataQiu commented:
/lgtm

pacoxu commented on content/en/docs/reference/setup-tools/kubeadm/kubeadm-init.md:
```diff
@@ -176,20 +176,24 @@ policy during rolling upgrades.
+`kubeadm.alpha.kubernetes.io/cri-socket` on the Node object. The new file is applied as an instance
+configuration patch, before any other user managed patches are applied when the `--patches` flag
+is used. It contains a single field `containerRuntimeEndpoint` from the
+[KubeletConfiguration file format](/docs/reference/config-api/kubelet-config.v1beta1/).
```
For upgrade, we will read from `kubeadm-flags.env` if `instance-config.yam` does not exist.

Should we mention it?

neolit123 replied:
adding:
> If the feature gate is enabled during upgrade, but the file `/var/lib/kubelet/instance-config.yaml` does not exists yet,
kubeadm will attempt to read the CRI socket value from the file `/var/lib/kubelet/kubeadm-flags.env`.

pacoxu reviewed (approved):
/lgtm

reylejano commented:
/approve
